File: ids-container-manager/src/main/java/de/fhg/ids/comm/unixsocket/TrustmeUnixSocketResponseHandler.java
/*-
 * ========================LICENSE_START=================================
 * IDS Container Manager
 * %%
 * Copyright (C) 2017 Fraunhofer AISEC
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =========================LICENSE_END==================================
 */
package de.fhg.ids.comm.unixsocket;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class TrustmeUnixSocketResponseHandler {
	private Logger LOG = LoggerFactory.getLogger(TrustmeUnixSocketResponseHandler.class);
	private byte[] rsp = null;
	
	public synchronized boolean handleResponse(byte[] rsp) {
		this.rsp = rsp.clone();
		this.notify();
		return true;
	}
	
	public synchronized byte[] waitForResponse() {
		while(this.rsp == null) {
			try {
				this.wait();
			} catch (InterruptedException e) {
				LOG.error(e.getMessage(),e);
			}
		}
		byte[] result = rsp;
		LOG.debug("received response byte length: {}", result.length);
		this.rsp = null;
		return result;
	}
}

File: ids-container-manager/src/main/kotlin/de/fhg/aisec/ids/comm/unixsocket/TrustmeUnixSocketResponseHandler.kt
/*-
 * ========================LICENSE_START=================================
 * ids-container-manager
 * %%
 * Copyright (C) 2019 Fraunhofer AISEC
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =========================LICENSE_END==================================
 */
package de.fhg.aisec.ids.comm.unixsocket

import org.slf4j.LoggerFactory
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

class TrustmeUnixSocketResponseHandler {
    private val lock = ReentrantLock()
    private val condition = lock.newCondition()

    private val LOG = LoggerFactory.getLogger(TrustmeUnixSocketResponseHandler::class.java)
    private var rsp: ByteArray? = null

    fun handleResponse(rsp: ByteArray): Boolean {
        lock.withLock {
            this.rsp = rsp.clone()
            condition.signal()
            return true
        }
    }

    fun waitForResponse(): ByteArray {
        lock.withLock {
            while (rsp == null) {
                try {
                    condition.await()
                } catch (e: InterruptedException) {
                    LOG.error(e.message, e)
                }
            }
            val result = rsp!!
            LOG.debug("received response byte length: {}", result.size)
            rsp = null
            return result
        }
    }
}
